Operator developer guidance

cccclai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72470



Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

Commit:
Operator developer guidance

Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"


Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"


Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+# Guidance for Operator Developer
```
I think Tugsuu is also adding this file in his PR:
https://github.com/pytorch/pytorch/pull/72313/files

Could you coordinate with him to land it? Thanks

cccclai replied:
Yeah @tugsbayasgalan  reached out to me and mentioned about it. Sorry should've checked with him so there isn't overlap. I will address the comments from you and update this pr.

cccclai replied:
https://github.com/pytorch/pytorch/pull/72313 includes code changes to skip bc check for operators with valid upgraders, as well as the guidance. It's probably better to have one dedicated pr for the guidance. I've added the changes from that pr to here and will address the feedback too.

tugsbayasgalan replied:
Yeah. I was updating our FC/BC checking test and I needed to reference this doc so I included my own version in my PR. But I think it makes sense to land this PR first

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+# Guidance for Operator Developer
```
We should have an intro on what the base concepts are: operator upgrader, BC and BC breaking, ...
Keep it short and simple.
You may already have something in the design review.

cccclai replied:
Added

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+
+2. Make changes to the operator and write an upgrader.
+    1. Make the operator change.
+    2. Write an upgrader in `caffe2/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp` file inside a map `kUpgradersEntryMap`. The softly enforced naming format is `&lt;operator_name>_&lt;operator_overload>_&lt;start>_&lt;end>`. For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need to be replaced by these upgraders.
```
will this render ok?

cccclai replied:
Resolve

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+The steps to write upgrader:
+
+1. Prepare a model before making changes to the operator, following the process:
+     1. add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that, an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following
```
```suggestion
     1. Add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following
```

raziel replied:
probably should be another step

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+The steps to write upgrader:
+
+1. Prepare a model before making changes to the operator, following the process:
+     1. add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that, an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following
```
what does this mean?

cccclai replied:
Add more explanation and hope it's clearer.

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+        ALL_MODULES = {
+           TestVersionedLinspaceV7(): "aten::linspace",
+        }
+        ```
````
Fix formatting for code blocks.

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+        python test/jit/fixtures_src/generate_models.py
+        ```
+
+       3. Commit the change and submit a pr
````
```suggestion
       3. Commit the change and submit a pull request.
```

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+
+The steps to write upgrader:
+
+1. Prepare a model before making changes to the operator, following the process:
```
Here you mean a test model, right?
Otherwise it's not clear what this model is mean to do or why. Please briefly explain.

cccclai replied:
Add the reason why a test model is needed.

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+    python pytorch/tools/codegen/operator_versions/gen_mobile_upgraders.py
+    ```
+
+    5. Add test:
````
Is this a test or tests?

raziel reviewed (commented):
I think this needs a bit more refinement.

Try to read it as if you knew nothing... or better as if it was code and by following every step you would actually end up with a proper upgrader PR.

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,95 @@
+
+2. Make changes to the operator and write an upgrader.
+    1. Make the operator change.
+    2. Write an upgrader in `caffe2/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp` file inside a map `kUpgradersEntryMap`. The softly enforced naming format is `&lt;operator_name>_&lt;operator_overload>_&lt;start>_&lt;end>`. For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need to be replaced by these upgraders.
```
Maybe we should explain what version is? Because for general public, it is not very clear

cccclai replied:
Add more explanation.

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+        }
+    ```
+
+    2. Bump the `kProducedFileFormatVersion` by 1 and provide the reasons under `caffe2/caffe2/versions.h `
````
I think you also need to bump `kMaxProducedFileFormatVersion`

cccclai replied:
Resolve

tugsbayasgalan commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+    constexpr uint64_t kProducedFileFormatVersion = 0x9L;
+    ```
+
+    3. In `caffe2/torch/csrc/jit/operator_upgraders/version_map.cpp`, add changes like below. You will need to make sure that the entry is **SORTED** according to the version bump number.
````
Maybe mention how the below example is formatted according to what we describe here

cccclai replied:
Added an example to explain

Commit:
Update on "Operator developer guidance"


Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

cccclai commented on torch/csrc/jit/operator_upgraders/README.md:
````diff
@@ -0,0 +1,95 @@
+        "aten::linspace(Scalar start, Scalar end, int? steps=None, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor"}}},
+    ```
+
+    4. After rebuild PyTorch, run the following command and it will auto generate a change to `caffe2/torch/csrc/jit/mobile/upgrader_mobile.cpp`. After rebuild PyTorch from source (`python setup.py`), run
````
1. "auto generate a change to upgrader_mobile.cpp" -> "auto update the file upgrader_mobile.cpp"
2. Add the link to build pytorch
3. Add the first time to build in step 1

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

iseeyuan commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,202 @@
+The steps to write upgrader:
+
+1. [Build PyTorch from souce](https://github.com/pytorch/pytorch#from-source) and prepare a test model before making changes to the operator, following the process below. A test model before making the operator changes is needed to test the upgrader. Otherwise, after the change to operator, the new runtime will no longer be able to produce a model with the historic operator and can't test it anym ...
+     1. Add a test module in `test/jit/fixtures_srcs/fixtures_src.py`. In `test/jit/fixtures_srcs/generate_models.py`, add a module
```
```suggestion
     1. Add a test module in `test/jit/fixtures_srcs/fixtures_src.py`. In `test/jit/fixtures_srcs/generate_models.py`, 
```

iseeyuan commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,202 @@
+
+The steps to write upgrader:
+
+1. [Build PyTorch from souce](https://github.com/pytorch/pytorch#from-source) and prepare a test model before making changes to the operator, following the process below. A test model before making the operator changes is needed to test the upgrader. Otherwise, after the change to operator, the new runtime will no longer be able to produce a model with the historic operator and can't test it anym ...
```
It's confusing for bullet points. The "big" first step of "build from sources" covers "small" steps of adding testing modules. It does not make sense to me. Some reformatting may be needed.

iseeyuan reviewed (approved):
Added some minor inline suggestions. LGTM in general. Thanks!

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

Commit:
Update on "Operator developer guidance"




Use markdown to write an upgrader guidance

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964/)

Differential Revision: [D34054964](https://our.internmc.facebook.com/intern/diff/D34054964)

[ghstack-poisoned]

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,205 @@
+# Guidance for Operator Developer
+
+PyTorch’s operators sometimes require changes to maintain the high quality user experience (UX) that PyTorch is known for. These changes can be backward compatibility (BC) breaking, where older programs will no longer run as expected on the latest version of PyTorch (an old writer / new reader problem) or forward compatibility (FC) breaking, where new programs will not run on older versions of Py ...
```
A few things:
* I would rephrase old write / new reader to actually talk about program and runtime.
* We talk about FC but we should clarify upgraders are just for BC>
*  "The replacement will only happen for old models, and it does not need to consider the new models." => Shouldn't this just be as simple as "upgraders only apply to instances of old operators to make sure they comply with the new operator contract"?

cccclai replied:
Address on the new pr

raziel commented on torch/csrc/jit/operator_upgraders/README.md:
```diff
@@ -0,0 +1,205 @@
+
+PyTorch’s operators sometimes require changes to maintain the high quality user experience (UX) that PyTorch is known for. These changes can be backward compatibility (BC) breaking, where older programs will no longer run as expected on the latest version of PyTorch (an old writer / new reader problem) or forward compatibility (FC) breaking, where new programs will not run on older versions of Py ...
+
+After you change to operator either the operator schema is BC-breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking iff they are used in TorchScript or mobile. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `
```
There's typos here: "After you change to operator either the operator schema is BC-breaking way or the semantics of the operator"

Also, this should at least include some examples on what's BC breaking.

"change non-BC breaking iff they are used in TorchScript or mobile"
Just for mobile? Also how could anybody know that: they can't so this is not optional.

cccclai replied:
Address on the new pr

raziel reviewed (commented):
I just reviewed the top since this is a closed PR and cannot make suggestions in the content itself better for me to continue my review once you start a new PR.

cccclai commented:
@raziel thanks for the feedback. I start a new pr https://github.com/pytorch/pytorch/pull/72818 to address the comments.
